[SYCL][CUDA] Implement Intel USM extension

fwyzard opened the pull request:
Implement the following functions in the CUDA plugin, and mark the tests for the USM features that are now supported.

Also, fix the CUDA version reported by SYCL.

#### Device
  - USM-related calls to `piDeviceGetInfo()`

#### Kernel
  - `piextKernelSetArgPointer()`

#### USM
  - `piextUSMHostAlloc()`
  - `piextUSMDeviceAlloc()`
  - `piextUSMSharedAlloc()`
  - `piextUSMFree()`
  - `piextUSMEnqueueMemset()`
  - `piextUSMEnqueueMemcpy()`
  - `piextUSMEnqueuePrefetch()`
  - `piextUSMEnqueueMemAdvise()` (see below)
  - `piextUSMGetMemAllocInfo()` (see below)

#### Limitations
As the Intel USM extension is still incomplete:
  - `piextUSMEnqueuePrefetch()` ignores the "flags" argument;
  - `piextUSMEnqueueMemAdvise()` does nothing.

fwyzard commented:
This PR does not mean to be exhaustive; it implements only the minimal part of the USM interface to get some test applications running.

bader commented:
@fwyzard, thanks a lot for working on this.
@Ruyk, @bjoernknafla, could you take a look, please?

fwyzard commented:
Updated to use `cuMemcpyAsync` directly, instead of dispatching to `cuMemcpyHtoDAsync`, `cuMemcpyDtoHAsync`, etc.

Ruyk commented on sycl/plugins/cuda/pi_cuda.cpp:
```diff
@@ -2863,6 +2871,130 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
+pi_result cuda_piextUSMHostAlloc(void **result_ptr, pi_context context,
+                                 pi_usm_mem_properties *properties, size_t size,
+                                 pi_uint32 alignment) {
+  PI_CHECK_ERROR(cuMemAllocHost(result_ptr, size));
```
Should there be an assert for properties to be nullptr since its not implemented?

Ruyk replied:
Alternatively, you could make this cuMemHostAlloc instead. We could map later the `CL_MEM_ALLOC_WRITE_COMBINED_INTEL` flag to `CU_MEMHOSTALLOC_WRITECOMBINED` assuming they are equivalent enough (they might not be)

fwyzard replied:
[cl_intel_unified_shared_memory.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc) says
> `CL_MEM_ALLOC_WRITE_COMBINED_INTEL`: Request write combined (WC) memory. Write combined memory may improve performance in some cases, however write combined memory must be used with care since it may hurt performance in other cases or use different coherency protocols than non-write combined memory.

which seems vague enough to be implementation-defined .

However, I don't understand how `properties` should be used.

I was expecting `properties` to point to a value of type `cl_mem_alloc_flags_intel`, which could have the possible values `CL_MEM_ALLOC_DEFAULT_INTEL` or `CL_MEM_ALLOC_WRITE_COMBINED_INTEL`.

Instead it points to a value of  type  `pi_usm_mem_properties`, which is a `typedef` for `_pi_usm_mem_properties`, which is a `typedef` for
```
enum : pi_uint64 {
  PI_MEM_ALLOC_FLAGS = CL_MEM_ALLOC_FLAGS_INTEL
} _pi_usm_mem_properties;
```
, where `CL_MEM_ALLOC_FLAGS_INTEL` is `#define`d as `0x4195`.

@jbrodman I see that you authored most of [cl_intel_unified_shared_memory.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc); could you clarify ?

bader replied:
Sorry, meant add @bashbaug to this conversation.

jbrodman replied:
We probably need to fill in some more PI values.  I hadn't captured them all when I did the PI USM rewrite.

bashbaug replied:
At least on the OpenCL side, the `properties` are intended to provide extra information for a USM allocation.  Right now, the only valid property name is `CL_MEM_ALLOC_FLAGS_INTEL` to specify a property value bitfield `cl_mem_alloc_flags_intel`, which means they are really only useful to request write combined memory.

We have had a request to support traditional [CL_MEM_FLAGS](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_API.html#CL_MEM_FLAGS), to provide information such as whether an allocation is read-only, or inaccessible from the host, so don't be too surprised if that shows up in the future.

Note that all of these properties are optional - they give the runtime more information about an allocation that could improve performance, but they aren't required for functional correctness.

fwyzard replied:
> Alternatively, you could make this cuMemHostAlloc instead. We could map later the `CL_MEM_ALLOC_WRITE_COMBINED_INTEL` flag to `CU_MEMHOSTALLOC_WRITECOMBINED` assuming they are equivalent enough (they might not be)

Allocating write-combined memory in CUDA has the disadvantage that doesn't allow using the same pointer on the host and the device any more. From [5.13. Unified Addressing](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__UNIFIED.html#group__CUDA__UNIFIED):
> ### Exceptions, Disjoint Addressing
>
>Not all memory may be accessed on devices through the same pointer value through which they are accessed on the host. These exceptions are host memory registered using `cuMemHostRegister()` and host memory allocated using the flag `CU_MEMHOSTALLOC_WRITECOMBINED`. For these exceptions, there exists a distinct host and device address for the memory. The device address is guaranteed to not overlap any valid host pointer range and is guaranteed to have the same value across all contexts that support unified addressing. 

Unless there is a mechanism to handle distinct host and device pointers in the SYCL runtime or in the USM extension, I don't think implementing write-combined memory is a good idea.

fwyzard replied:
> Should there be an assert for properties to be nullptr since its not implemented?

I have to learn how properties work :-)

fwyzard replied:
OK, understood.

Currently, all calls to `piextUSMHostAlloc`, `piextUSMDeviceAlloc` and `piextUSMSharedAlloc` in [sycl/source/detail/usm/usm_impl.cpp](https://github.com/intel/llvm/blob/dc729a72b5bd815ffd7e118cd8d1992c5248ee4a/sycl/source/detail/usm/usm_impl.cpp) pass `nullptr` in the properties, so I've added an `assert` in that regard.

Ruyk commented on sycl/plugins/cuda/pi_cuda.cpp:
```diff
@@ -2863,6 +2871,130 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
+    retErr = PI_CHECK_ERROR(cuMemFree((CUdeviceptr)ptr));
+  }
+  if (type == CU_MEMORYTYPE_HOST) {
+    retErr = PI_CHECK_ERROR(cuMemFreeHost(ptr));
```
[`cuMemAllocHost`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1gdd8311286d2c2691605362c689bc64e0)  doesn't require `cuMemFreeHost`, only [`cuMemHostAlloc`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1g572ca4011bfcb25034888a14d4e035b9) does.

fwyzard replied:
The documentation of [cuMemFreeHost](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1g62e0fdbe181dab6b1c90fa1a51c7b92c) says

> Frees the memory space pointed to by `p`, which must have been returned by a previous call to `cuMemAllocHost()`.

Ruyk commented on sycl/plugins/cuda/pi_cuda.cpp:
```diff
@@ -2863,6 +2871,130 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
+  unsigned int type;
+  retErr = PI_CHECK_ERROR(cuPointerGetAttribute(
+      &type, CU_POINTER_ATTRIBUTE_MEMORY_TYPE, (CUdeviceptr)ptr));
+  assert(type == CU_MEMORYTYPE_DEVICE || type == CU_MEMORYTYPE_HOST);
```
Seems to me an assert is not enough in this case, this should be a proper PI_INVALID_VALUE error returned at runtime in case is none of the two.

fwyzard replied:
According to the documentation for [for `cuPointerGetAttribute`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__UNIFIED.html#group__CUDA__UNIFIED_1g0c28ed0aff848042bc0533110e45820c) these should be the only possible values:

> #### `CU_POINTER_ATTRIBUTE_MEMORY_TYPE`:
> 
> Returns in `*data` the physical memory type of the memory that `ptr` addresses as a `CUmemorytype` enumerated value. The type of data must be `unsigned int`.
> 
> If `ptr` addresses device memory then `*data` is set to `CU_MEMORYTYPE_DEVICE`. The particular `CUdevice` on which the memory resides is the `CUdevice` of the `CUcontext` returned by the `CU_POINTER_ATTRIBUTE_CONTEXT` attribute of `ptr`.
> 
> If `ptr` addresses host memory then `*data` is set to `CU_MEMORYTYPE_HOST`.
> 
> If `ptr` was not allocated by, mapped by, or registered with a `CUcontext` which uses unified virtual addressing then `CUDA_ERROR_INVALID_VALUE` is returned.
> 
> If the current `CUcontext` does not support unified virtual addressing then `CUDA_ERROR_INVALID_CONTEXT` is returned.

On the other hand, the documentation for [`enum CUmemorytype`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1g8a114cc994ad2e865c44ef3838eaec72) does list the additional values `CU_MEMORYTYPE_ARRAY` and `CU_MEMORYTYPE_UNIFIED`, so I can look into it a bit more.

jbrodman replied:
I think the test for shared will involve CU_POINTER_ATTRIBUTE_IS_MANAGED

fwyzard replied:
here's what I got from an explicit test:
```
Found 1 CUDA devices
Using device GeForce GTX 960M
cuMemAllocHost() returns a pointer of type CU_MEMORYTYPE_HOST
cuMemHostAlloc() returns a pointer of type CU_MEMORYTYPE_HOST
cuMemAlloc() returns a pointer of type CU_MEMORYTYPE_DEVICE
cuMemAllocManaged() returns a pointer of type CU_MEMORYTYPE_DEVICE
```

fwyzard replied:
which helps, since the memory returned by `cuMemAllocManaged()` should be freed by `cuMemFree()` as well.

Ruyk replied:
there is also [`cuMemHostRegister`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1gf0a9fe11544326dabd743b7aa6b54223), which from what I can _infer_ from the docs, returns `CU_MEMORYTYPE_HOST`, but I would expect `CU_MEMORYTYPE_UNIFIED`. I find that a bit confusing.
I guess an assert is OK for this, if the other values of the struct are not returned.

jbrodman replied:
I think cuMemHostRegister only pins and maps existing c/c++ allocations so they can be visible like "host" allocations.

fwyzard replied:
OK, I tested it and
```
cuMemHostRegister() returns a pointer of type CU_MEMORYTYPE_HOST
```

However, that pointer **cannot** be freed by `cuMemFreeHost`:
```
cudaCheck(cuMemFreeHost(buffer));
CUDA_ERROR_INVALID_VALUE: invalid argument
```

fwyzard replied:
>  I guess an assert is OK for this, if the other values of the struct are not returned.

OK, I'll switch back to an assert.

jbrodman replied:
Correct - this is memory that _was not allocated by cuda_.

The programmer should `cuMemHostUnregister` then deallocate with either `std::free` or `delete`.

fwyzard commented on sycl/plugins/cuda/pi_cuda.cpp:
```diff
@@ -2863,6 +2871,130 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
+  if (type == CU_MEMORYTYPE_HOST) {
+    retErr = PI_CHECK_ERROR(cuMemFreeHost(ptr));
+  }
+  return retErr;
```
Understood. I will adopt the same pattern for all functions that use `PI_CHECK_ERROR`.

jbrodman commented on sycl/plugins/cuda/pi_cuda.cpp:
```diff
@@ -2863,6 +2871,130 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
+                                   pi_device device,
+                                   pi_usm_mem_properties *properties,
+                                   size_t size, pi_uint32 alignment) {
+  cl::sycl::detail::pi::die("cuda_piextUSMSharedAlloc not implemented");
```
auto Err = cuMemAllocManaged(/*pointer goes here */, Size, CU_MEM_ATTACH_GLOBAL);

fwyzard replied:
Implemented.

bjoernknafla commented on sycl/plugins/cuda/pi_cuda.cpp:
```diff
@@ -2863,6 +2871,130 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
+                                   pi_device device,
+                                   pi_usm_mem_properties *properties,
+                                   size_t size, pi_uint32 alignment) {
+  assert(device != nullptr);
```
```cpp
assert(result_ptr != nullptr);
assert(context != nullptr);
```

bjoernknafla commented on sycl/plugins/cuda/pi_cuda.cpp:
```diff
@@ -2863,6 +2871,130 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
+                                   size_t size, pi_uint32 alignment) {
+  assert(device != nullptr);
+  ScopedContext active(context);
+  PI_CHECK_ERROR(cuMemAlloc((CUdeviceptr *)result_ptr, size));
```
try-catch-block currently (sigh) required around `PI_CHECK_ERROR` to translate it into return codes.

bader commented:
+@bashbaug

fwyzard commented on sycl/plugins/cuda/pi_cuda.cpp:
```diff
@@ -2863,6 +2871,130 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
+}
+
+pi_result cuda_piextUSMFree(pi_context context, void *ptr) {
+  pi_result retErr = PI_SUCCESS;
```
is it allowed to pass a `nullptr` context for a host pointer ?

Ruyk replied:
I don't think the SYCL RT gets to this call on the host device, it should be an assert for that case.

fwyzard replied:
In fact I hope is does pass a valid context, because it turns out that `cuMemFreeHost()` needs a valid CUDA context.

fwyzard commented:
I've implemented `piextUSMSharedAlloc()` and `piextUSMEnqueueMemset()`, and hopefully addressed most of the comments.

I still need to look into using `cuMemHostAlloc()` with the optional write-combine flag instead of `cuMemAllocHost()`.

If I missed anything, please mention it (again).

Ruyk commented:
Still looking to see if there are more lit tests passing, USM tests are marked as "unsupported" for the CUDA backend so they won't report if they are passing now.

fwyzard commented:
> Still looking to see if there are more lit tests passing, USM tests are marked as "unsupported" for the CUDA backend so they won't report if they are passing now.

How do I run the USM tests with the CUDA backend ?

I have tried removing the `// XFAIL: cuda` line from all files under `sycl/test/usm/`, and running `make check-sycl-cuda-usm` in the build directory, and I got:
```
...
getDeviceCount cpu:PI_CUDA
Found available CPU device
getDeviceCount gpu:PI_CUDA
Found available GPU device
getDeviceCount accelerator:PI_CUDA
Could not find AOT device compiler opencl-aot
Found AOT device compiler ocloc
Could not find AOT device compiler aoc

Testing Time: 14.15s
  Expected Passes    : 24
```

Which is not what I expected, because
  - "Found available CPU device" makes no sense for CUDA
  - "Expected Passes    : 24" (i.e. all) is too many, I would have expected same failers, since not all USM functions are implements.

fwyzard commented:
In fact, looking at the tests being built and run, it looks like `make check-sycl-cuda-usm`
  - still builds with `-fsycl-targets=spir64-unknown-linux-sycldevice` instead of `-fsycl-targets=nvptx64-unknown-linux-sycldevice`
  - does not pass `SYCL_BE=PI_CUDA` at runtime

fwyzard commented:
I have built and run by hand the tests with the CUDA backend, and removed the XFAIL from the ones that worked.

fwyzard commented:
I've added an implementation for the USM-related calls to `piDeviceGetInfo()`, based on my understanding of what they are supposed to query.

The documentation of the extension is not very clear in this regard, so I had to mostly guess...

jbrodman commented:
> I've added an implementation for the USM-related calls to `piDeviceGetInfo()`, based on my understanding of what they are supposed to query.
> 
> The documentation of the extension is not very clear in this regard, so I had to mostly guess...

Get pointer type should accept a pointer and return 1 of 4 values:
unknown: cuda/sycl knows nothing about this pointer, so it was probably from malloc
host: ptr came from cuHostAlloc or whatever
device: ptr came from cuMalloc
shared: ptr came from cuMallocManaged

get_pointer_device should return the sycl device that the pointer was allocated against (for device/shared) pointers

fwyzard commented:
Thanks for the initiation.

Actually I meant about `PI_USM_SINGLE_SHARED_SUPPORT` and similar.

jbrodman commented:
> Thanks for the initiation.
> 
> Actually I meant about `PI_USM_SINGLE_SHARED_SUPPORT` and similar.

Ah - yes.
That's actually more about a device specific behavior.  It's not in the SYCL part yet, only the CL one, but it will probably show up in a few months.

fwyzard commented:
> That's actually more about a device specific behavior. It's not in the SYCL part yet, only the CL one, but it will probably show up in a few months.

I'm confused :-(

I added them because the USM LIT tests were failing on the CUDA backend with errors like
```
$ SYCL_BE=PI_CUDA ./memcpy.cpp.out
pi_die: Unknown parameter 16786 passed to cuda_piDeviceGetInfo

terminate called without an active exception
Aborted (core dumped)
```
The error is caused by https://github.com/intel/llvm/blob/9134bfa7f96fe643dc17165d046724873d48b1a1/sycl/test/usm/mixed_queue.cpp#L25-L28
it looks like `device::get_info<N>()` maps to `get_device_info<bool,  N>::get(device, plugin)` which calls `piDeviceGetInfo()`; so the tests actually exercise those device queries.

The OpenCL plugin maps `piDeviceGetInfo()` directly to `clGetDeviceInfo()`.

In the CUDA plugin tried to query the CUDA backend and return the equivalent information.

What it's not clear from the [cl_intel_unified_shared_memory.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc) extension is if all value of the `cl_device_unified_shared_memory_capabilities_intel` bit field make sense for all calls to `clGetDeviceInfo(...MEM_CAPABILITIES_INTEL)`, and how they map to the different USM levels (Explicit, Restricted, Concurrent,  System).

fwyzard commented:
> What it's not clear from the [cl_intel_unified_shared_memory.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc) extension is if all value of the `cl_device_unified_shared_memory_capabilities_intel` bit field make sense for all calls to `clGetDeviceInfo(...MEM_CAPABILITIES_INTEL)`, and how they map to the different USM levels (Explicit, Restricted, Concurrent, System).

For example, what is a call to `device.get_info<info::device::usm_host_allocations>()` supposed to check ?

Whether the `device` can do any of
   - may access (read or write) Unified Shared Memory allocated on the host
   - may perform atomic memory operations on Unified Shared Memory on the host
   - may access Unified Shared Memory allocated on the host, concurrently with the host or an other device
   - may perform atomic memory operations on Unified Shared Memory on the host, concurrently with the host or an other device

?
In this case, what would "Unified Shared Memory allocated on the host" mean ?
Memory initially allocated with `clHostMemAllocINTEL` ? Or with `clSharedMemAllocINTEL` ?

fwyzard commented:
This PR is starting to be fairly complete, though it is probably not robust.

The main limitation comes for the lack of documentation of the corner case, and/or examples of how the USM calls are supposed to interact with each other.

fwyzard commented:
@bashbaug @jbrodman

About the behaviour of `piDeviceGetInfo()` for `PI_USM_HOST_SUPPORT`, `PI_USM_DEVICE_SUPPORT`, `PI_USM_SINGLE_SHARED_SUPPORT`, `PI_USM_CROSS_SHARED_SUPPORT`, and `PI_USM_SYSTEM_SHARED_SUPPORT`: re-reading the [OpenCL cl_intel_unified_shared_memory extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc), the [SYCL USM proposal](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/USM.adoc), the [Level Zero Memory documentation](https://spec.oneapi.com/versions/0.6.0/oneL0/core/PROG.html#linearmemory) and the [CUDA driver API for Unified Addressing](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__UNIFIED.html#group__CUDA__UNIFIED), it's starting to make sense...

However, it looks like there are still some corner case that are not very clear.
Is the API finalised ?

According to Table 5 in [cl_intel_unified_shared_memory](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc):

>  The host memory access capabilities apply to any host allocation.

so, `PI_USM_HOST_SUPPORT` queries whether a CUDA device can access memory allocated on the host by `cuMemAllocHost` (or `cuMemHostAlloc`), i.e. *zero-copy* memory with Unified Addressing.

---

>  The device memory access capabilities apply to any device allocation associated with this device.

so `PI_USM_DEVICE_SUPPORT` queries whether a CUDA device can access memory allocated on the device *itself* with `cuMemAlloc`, which I would take for granted ?

The table under [USM Allocations](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/USM.adoc#usm-allocations) in the USM proposal mentions for the `device` allocations also the possibility of peer-to-peer access from other devices. How is that queried ?
`piDeviceGetInfo()` has a single `device` parameter, so I wouldn't know how to check that.

---

>  The single device shared memory access capabilities apply to any shared allocation associated with this device.

so `PI_USM_SINGLE_SHARED_SUPPORT`  query whether the device supports managed memory, allocated with `cuMemAllocManaged`.

---

>  The cross-device shared memory access capabilities apply to any shared allocation associated with this device, or to any shared memory allocation on another device that also supports the same cross-device shared memory access capability.

so `PI_USM_CROSS_SHARED_SUPPORT` queries whther the device supports peer-to-peer access to managed memory.

---

>  The shared system memory access capabilities apply to any allocations made by a system allocator, such as `malloc` or `new`.

so `PI_USM_SYSTEM_SHARED_SUPPORT` queries whether the device can access host memory allocated by the system allocator (e.g. `malloc()` and `new`); for CUDA devices this is only available on Power9 machines as far as I know, so I have no way to test it.

fwyzard commented:
rebased

jbrodman commented:
> @bashbaug @jbrodman
> 
> About the behaviour of `piDeviceGetInfo()` for `PI_USM_HOST_SUPPORT`, `PI_USM_DEVICE_SUPPORT`, `PI_USM_SINGLE_SHARED_SUPPORT`, `PI_USM_CROSS_SHARED_SUPPORT`, and `PI_USM_SYSTEM_SHARED_SUPPORT`: re-reading the [OpenCL cl_intel_unified_shared_memory extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc), the [SYCL USM proposal](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/USM.adoc), the [Level Zero Memory documentation](https://spec.oneapi.com/versions/0.6.0/oneL0/core/PROG.html#linearmemory) and the [CUDA driver API for Unified Addressing](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__UNIFIED.html#group__CUDA__UNIFIED), it's starting to make sense...
> 
> However, it looks like there are still some corner case that are not very clear.
> Is the API finalised ?

No.  It's mostly complete, but there are still things that might change as we get more experience and usage.

> According to Table 5 in [cl_intel_unified_shared_memory](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc):
> 
> > The host memory access capabilities apply to any host allocation.
> 
> so, `PI_USM_HOST_SUPPORT` queries whether a CUDA device can access memory allocated on the host by `cuMemAllocHost` (or `cuMemHostAlloc`), i.e. _zero-copy_ memory with Unified Addressing.


Yes.

> > The device memory access capabilities apply to any device allocation associated with this device.
> 
> so `PI_USM_DEVICE_SUPPORT` queries whether a CUDA device can access memory allocated on the device _itself_ with `cuMemAlloc`, which I would take for granted ?

Yes.

> 
> The table under [USM Allocations](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/USM.adoc#usm-allocations) in the USM proposal mentions for the `device` allocations also the possibility of peer-to-peer access from other devices. How is that queried ?

TBD.


> > The single device shared memory access capabilities apply to any shared allocation associated with this device.
> 
> so `PI_USM_SINGLE_SHARED_SUPPORT` query whether the device supports managed memory, allocated with `cuMemAllocManaged`.
> 

Correct.

> > The cross-device shared memory access capabilities apply to any shared allocation associated with this device, or to any shared memory allocation on another device that also supports the same cross-device shared memory access capability.
> 
> so `PI_USM_CROSS_SHARED_SUPPORT` queries whther the device supports peer-to-peer access to managed memory.
>

Not entirely.  P2P is kind of a special case.  This is more is one shared allocation able to migrate between different devices in a platform.  P2P is really more about if it can migrate *directly* without needing to go through the host.

> > The shared system memory access capabilities apply to any allocations made by a system allocator, such as `malloc` or `new`.
> 
> so `PI_USM_SYSTEM_SHARED_SUPPORT` queries whether the device can access host memory allocated by the system allocator (e.g. `malloc()` and `new`); for CUDA devices this is only available on Power9 machines as far as I know, so I have no way to test it.

Correct.  I don't expect to really be done until multiple vendors implement the same protocol for coherent device attach.

fwyzard commented:
@jbrodman thank you very much for the clarifications.

> > `PI_USM_SINGLE_SHARED_SUPPORT` query whether the device supports managed memory, allocated with `cuMemAllocManaged`.
> 
> Correct.
>
> > `PI_USM_CROSS_SHARED_SUPPORT` queries whether the device supports peer-to-peer access to managed memory.
> 
> Not entirely. P2P is kind of a special case. This is more one shared allocation able to migrate between different devices in a platform. P2P is really more about if it can migrate _directly_ without needing to go through the host.

I see.
Then I think I should slightly change the answer for the queries to `PI_USM_SINGLE_SHARED_SUPPORT` and `PI_USM_CROSS_SHARED_SUPPORT` to reflect this.

bader commented:
@fwyzard the latest commit is not signed and it looks like it have been squashed.
Could you also update the branch, please? I merged conflicting PR: https://github.com/intel/llvm/pull/1299.

Other than that I don't have any other comments to this patch. Thanks!

fwyzard commented:
Fixed commit history, rebased on top of #1310 .

bjoernknafla commented:
https://github.com/intel/llvm/pull/1310 has been merged.

@fwyzard did you manage to get the LIT tests to work for CUDA? I am not sure (haven’t tested) if [your `SYCL_BE` env var fix](https://github.com/intel/llvm/pull/1288) would be enough or if you need the more extensive fix of the [`get-device-count-by-type` tool](https://github.com/intel/llvm/pull/1300) for it?

fwyzard commented:
> @fwyzard did you manage to get the LIT tests to work for CUDA? I am not sure (haven’t tested) if [your `SYCL_BE` env var fix](https://github.com/intel/llvm/pull/1288) would be enough or if you need the more extensive fix of the [`get-device-count-by-type` tool](https://github.com/intel/llvm/pull/1300) for it?

I'm actually testing with all these PRs on top o fthe `sycl` branch:
  - #1252 [SYCL] Fixes for multiple backends in the same program
  - #1288 [SYCL] Run the LIT tests using the selected backend
  - #1293 [SYCL][CUDA] Improve CUDA backend documentation
  - #1300 [SYCL][CUDA] Fix LIT testing with CUDA devices
  - #1302 [SYCL][CUDA] Replace assert with CHECK
  - #1303 [SYCL][CUDA] LIT XFAIL/UNSUPPORTED
  - #1304 [SYCL][CUDA] Lit exceptions

but I haven't gotten around to trying the LIT tests, yet.
